File: scripts/check_case_links.py
#!/usr/bin/env python3
"""
Regression guard for the Academic Misconduct Dilemma App's source links.
Verifies that all caseStudy.url and standardUrl fields in misconduct-data.js
are present and point to specific direct sources rather than generic search/index pages.
"""

import json
import re
import sys
from pathlib import Path

data_path = Path(__file__).resolve().parent.parent / "misconduct-data.js"
if not data_path.exists():
    print(f"FAIL: {data_path} not found")
    sys.exit(1)

content = data_path.read_text(encoding="utf-8")
raw = content.split("=", 1)[1].rsplit(";", 1)[0].strip()

try:
    data = json.loads(raw)
except Exception as e:
    print(f"FAIL: could not parse misconduct-data.js: {e}")
    sys.exit(1)

BAD_PATTERNS = [
    (re.compile(r"\?s=", re.I), "generic search-query URL"),
    (re.compile(r"^https?://(www\.)?retractionwatch\.com/?$", re.I), "links to RetractionWatch homepage"),
    (re.compile(r"^https?://ori\.hhs\.gov/case_summary/?$", re.I), "links to generic ORI index"),
    (re.compile(r"^https?://(www\.)?google\.[a-z.]+/search", re.I), "links to search engine results page"),
    (re.compile(r"^http://blogs\.nature\.com/", re.I), "links to retired Nature blog domain (use nature.com/articles/ instead)")
]

failures = []

for item in data:
    fields = [
        ("caseStudy.url", item.get("caseStudy", {}).get("url")),
        ("standardUrl", item.get("standardUrl")),
    ]
    for field_name, url in fields:
        if not url:
            failures.append(f"{item['id']}: {field_name} is missing")
            continue
        for pat, reason in BAD_PATTERNS:
            if pat.search(url):
                failures.append(f"{item['id']}: {field_name} = \"{url}\" -> {reason}")

if failures:
    print(f"FAIL: {len(failures)} link-quality issue(s) found:")
    for f in failures:
        print(f" - {f}")
    sys.exit(1)

print(f"OK: all {len(data)} entries ({len(data)*2} URLs) have specific caseStudy.url and standardUrl.")
sys.exit(0)

File: scripts/check-case-links.js
#!/usr/bin/env node
/**
 * Regression guard for the Academic Misconduct Dilemma App's source links.
 *
 * Root cause this guards against: caseStudy.url / standardUrl fields that
 * point to a generic search-results or index page (e.g.
 * "https://retractionwatch.com/?s=Some+Name" or
 * "https://ori.hhs.gov/case_summary") instead of a specific, direct article
 * about the case. Those render as a working hyperlink in the UI, so the bug
 * is invisible to a quick smoke test -- clicking through is the only way to
 * notice the link doesn't land on the actual case.
 *
 * This script loads misconduct-data.js exactly the way the app does (it sets
 * window.MISCONDUCT_DATA) and flags any entry whose caseStudy.url or
 * standardUrl matches a known "not a specific case" pattern, or is missing.
 *
 * Usage: node scripts/check-case-links.js
 * Exit code 0 = pass, 1 = fail (with a list of offending entries).
 */

const path = require("path");

global.window = global.window || {};
require(path.join(__dirname, "..", "misconduct-data.js"));

const data = global.window.MISCONDUCT_DATA;

if (!Array.isArray(data) || data.length === 0) {
  console.error("FAIL: window.MISCONDUCT_DATA did not load or is empty.");
  process.exit(1);
}

// Patterns that indicate a link goes "somewhere" (a search box or a generic
// index) rather than to the specific case/article.
const BAD_PATTERNS = [
  {
    re: /\?s=/i,
    reason:
      "generic search-query URL (returns a results page whose content can change or come up empty, not a specific case)",
  },
  {
    re: /^https?:\/\/(www\.)?retractionwatch\.com\/?$/i,
    reason: "links to the RetractionWatch homepage, not a case",
  },
  {
    re: /^https?:\/\/ori\.hhs\.gov\/case_summary\/?$/i,
    reason: "links to the generic ORI case-summary index, not a specific finding",
  },
  {
    re: /^https?:\/\/(www\.)?google\.[a-z.]+\/search/i,
    reason: "links to a search engine results page",
  },
];

const failures = [];

for (const item of data) {
  const fields = [
    ["caseStudy.url", item.caseStudy && item.caseStudy.url],
    ["standardUrl", item.standardUrl],
  ];

  for (const [field, url] of fields) {
    if (!url) {
      failures.push(`${item.id}: ${field} is missing`);
      continue;
    }
    for (const { re, reason } of BAD_PATTERNS) {
      if (re.test(url)) {
        failures.push(`${item.id}: ${field} = "${url}" -> ${reason}`);
      }
    }
  }
}

if (failures.length) {
  console.error(`FAIL: ${failures.length} link-quality issue(s) found in misconduct-data.js:\n`);
  failures.forEach((f) => console.error(" - " + f));
  process.exit(1);
}

console.log(
  `OK: all ${data.length} entries have a specific caseStudy.url and standardUrl (no generic search or index pages).`
);
